### PR TITLE
Fixed lockfile check when using a Docker read-only filesystem

### PR DESCRIPTION
**Pull request by jfennick**
Fixes #

Changes made in this Pull Request:
 - Added OSError in addition to PermissionError

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by IAlibay** [comment]
Thanks @jfennick - since there are no related issues and it'll be hard to test the error, could you explain in more details what was the cause for this fix / any code/docker env details to duplicate the issue?

**Comment by jfennick** [comment]
Sure. Here is the relevant command line + stack trace from my log file.
```
/Users/jakefennick/workflow_inference_compiler/cachedir/fe47d299645b0f71eff2e8cbf791dcf9$ docker \
    run \
    -i \
    --mount=type=bind,source=/Users/jakefennick/workflow_inference_compiler/cachedir/fe47d299645b0f71eff2e8cbf791dcf9,target=/RvczZO \
    --mount=type=bind,source=/private/tmp/docker_tmpyrn_kcgl,target=/tmp \
    --mount=type=bind,source=/Users/jakefennick/workflow_inference_compiler/cachedir/1b588cb52801df5be9d783bad2d2cb63/npt.gro,target=/var/lib/cwl/stgdce8320f-73ef-4a3a-8db9-ea43afcfd074/npt.gro,readonly \
    --mount=type=bind,source=/Users/jakefennick/workflow_inference_compiler/cachedir/ae541201128048b4defdf1875c5dfa42/prod.trr,target=/var/lib/cwl/stg7d3f8150-98a4-4051-aed6-3f480f6cf22a/prod.trr,readonly \
    --workdir=/RvczZO \
    --read-only=true \
    --user=502:20 \
    --rm \
    --cidfile=/private/tmp/docker_tmp_n10lmap/20220916115735-218731.cid \
    --env=TMPDIR=/tmp \
    --env=HOME=/RvczZO \
    jakefennick/scripts \
    python3 \
    /align_protein_CA_mda.py \
    /var/lib/cwl/stgdce8320f-73ef-4a3a-8db9-ea43afcfd074/npt.gro \
    /var/lib/cwl/stg7d3f8150-98a4-4051-aed6-3f480f6cf22a/prod.trr \
    prod_align_protein_CA.trr
<Universe with 40893 atoms>
Traceback (most recent call last):
  File "/align_protein_CA_mda.py", line 15, in <module>
    trr.trajectory = TRR.TRRReader(input_trr_path) # This loads all frames.
  File "/miniconda/lib/python3.8/site-packages/MDAnalysis/lib/util.py", line 2495, in wrapper
    return func(self, *args, **kwargs)
  File "/miniconda/lib/python3.8/site-packages/MDAnalysis/coordinates/XDR.py", line 157, in __init__
    self._load_offsets()
  File "/miniconda/lib/python3.8/site-packages/MDAnalysis/coordinates/XDR.py", line 198, in _load_offsets
    with fasteners.InterProcessLock(lock_name) as filelock:
  File "/miniconda/lib/python3.8/site-packages/fasteners/process_lock.py", line 157, in __enter__
    gotten = self.acquire()
  File "/miniconda/lib/python3.8/site-packages/fasteners/process_lock.py", line 135, in acquire
    self._do_open()
  File "/miniconda/lib/python3.8/site-packages/fasteners/process_lock.py", line 107, in _do_open
    self.lockfile = open(self.path, 'a')
OSError: [Errno 30] Read-only file system: b'/var/lib/cwl/stg7d3f8150-98a4-4051-aed6-3f480f6cf22a/.prod.trr_offsets.lock'
```
For purposes of reproducibility/distribution with Common Workflow Language scripts, I am embedding [align_protein_CA_mda.py](https://github.com/jfennick/workflow_inference_compiler/blob/master/examples/scripts/align_protein_CA_mda.py) within my Docker image [jakefennick/scripts](https://hub.docker.com/repository/docker/jakefennick/scripts). My script works fine outside of Docker, but a Docker image is of course a read-only filesystem which causes MDAnalysis to raise an OSError, not a PermissionError.

For testing purposes, align_protein_CA_mda.py is rather trivial and should work with any standard gro and trr files.

**Comment by jfennick** [comment]
I guess I should clarify that `cwltool` automatically generates the above command and it includes the `--read-only=true` flag.

File: package/MDAnalysis/coordinates/XDR.py
Comment (tylerjereddy) [inline package/MDAnalysis/coordinates/XDR.py]: A reproducer for the original handled PermissionError is straightforward, like below, I wonder if we could cook something up for the other error? May not be worth much energy though.

```diff
--- a/testsuite/MDAnalysisTests/coordinates/test_xdr.py
+++ b/testsuite/MDAnalysisTests/coordinates/test_xdr.py
@@ -904,3 +904,14 @@ class TestTRRReader_offsets(_GromacsReader_offsets):
         9155712, 10300176
     ])
     _reader = mda.coordinates.TRR.TRRReader
+
+
+def test_offset_readonly_handle(tmpdir):
+    with tmpdir.as_cwd():
+        curr_path = os.getcwd()
+        new_gro = os.path.join(curr_path, os.path.basename(GRO))
+        new_xtc = os.path.join(curr_path, os.path.basename(XTC))
+        shutil.copyfile(GRO, new_gro)
+        shutil.copyfile(XTC, new_xtc)
+        os.chmod(curr_path, 0o555)
+        u = mda.Universe(new_gro, new_xtc)

```

Comment (zemanj) [inline package/MDAnalysis/coordinates/XDR.py]: According to the [Python docs](https://docs.python.org/3/library/exceptions.html#os-exceptions), `PermissionError` is a subclass of `OSError`, so your code is currently equivalent to only checking for `OSError`.
I'd advise against a simple check for `OSError` only, though, as this exception class covers many kinds of OS-related errors.
To narrow down the cause of the exception, it might be better to catch `OSError`, then check if it's a `PermissionError` **or** the `errno` member of the exception is the one you encountered. If so, raise the warning as before, otherwise just `raise` for the exception to bubble up.
For checking the errno, there should be a suitable constant in the [`errno` module of the standard library](https://docs.python.org/3/library/errno.html) so you don't have to hard-code numbers here.

EDIT: It's `errno.EROFS`.

**Review by zemanj — CHANGES_REQUESTED** [review]
Checking for `OSError` only seems too broad to me (for details see my code comment).

File: package/MDAnalysis/coordinates/XDR.py
Comment (zemanj) [inline package/MDAnalysis/coordinates/XDR.py:205]:
```suggestion
            else:
```

File: package/MDAnalysis/coordinates/XDR.py
Comment (zemanj) [inline package/MDAnalysis/coordinates/XDR.py]:
```suggestion
                raise
```

**Review by zemanj — DISMISSED** [review]
Awesome, just a few nitpicks. If you re-raise an exception, make sure to only use a bare `raise` statement without arguments. If you use `raise e`, the original exception trace won't be preserved.

File: package/MDAnalysis/coordinates/XDR.py
Comment (jfennick) [reply package/MDAnalysis/coordinates/XDR.py:205]: I'm assuming you didn't mean to remove the return statement. ?

File: package/MDAnalysis/coordinates/XDR.py
Comment (jfennick) [reply package/MDAnalysis/coordinates/XDR.py]: Thanks, I wasn't aware of that.

File: package/MDAnalysis/coordinates/XDR.py
Comment (zemanj) [reply package/MDAnalysis/coordinates/XDR.py:205]: You're absolutely right, that was bogus

**Review by zemanj — COMMENTED** [review]
Looks good to me. @tylerjereddy already commented on testing this so I'll approve once he's ok with it.

**Review by IAlibay — DISMISSED** [review]
I'm going to put a short hold on new contributions for a few days re: licensing (sorry about this) - I'll release a blog post explaining why tonight / tomorrow hopefully, nothing big just need to work out what's our plan before we continue

**Comment by jfennick** [comment]
Any movement on this issue?

**Comment by IAlibay** [comment]
Sorry about this @jfennick - I didn't get much headway on the alluded to blog post (fingers crossed end of this week). I'm dismissing my block, but could I ask you that you introduce yourself on the mailing list as per https://github.com/MDAnalysis/mdanalysis/pull/3832#pullrequestreview-1111137578 before we merge this?

We will need to trace contributors in the near future, so we need a way to reach out to everyone.

**Comment by jfennick** [comment]
Is there anything else I need to do to get this merged? If licensing issues are impeding drive-by contributions, then in the future perhaps I'll simply submit Issues instead of PRs.

**Comment by IAlibay** [comment]
@zemanj you have a blocking review here, could you have another look please?

**Comment by IAlibay** [comment]
> Is there anything else I need to do to get this merged? If licensing issues are impeding drive-by contributions, then in the future perhaps I'll simply submit Issues instead of PRs.

Sorry about the long time to merge, the licensing stuff is not a blocker here (I dismissed my review some time back alongside https://github.com/MDAnalysis/mdanalysis/pull/3832#issuecomment-1266267258). I'll give @zemanj until next week to re-review otherwise I'll take over and approve.

**Comment by hmacdope** [comment]
@IAlibay just scrolling through and thought I would give this a bump.

**Review by IAlibay — APPROVED** [review]
lgtm sorry this took so long @jfennick I'm not familiar with this part of the code so I didn't want to dismiss things rapidly

**Review by IAlibay — CHANGES_REQUESTED** [review]
Actually, I completely forgot - could you add yourself to `AUTHORS` @jfennick ?

**Review by IAlibay — APPROVED** [review]

**Comment by IAlibay** [comment]
Sorry about the last two commits, just thought it was faster for me to just add you to the changelog and fix the merge conflicts in one.

**Comment by jfennick** [comment]
> Sorry about the last two commits, just thought it was faster for me to just add you to the changelog and fix the merge conflicts in one.

No problem. Whatever works.

**Comment by IAlibay** [comment]
CI failure is being solved in #3886 I'm going to go ahead with the squash merge to not hold things up.

**Pull request merged**
